### PR TITLE
Add versioned API docs workflow

### LINKED ISSUE #4593: Versioned API documentation: proposal
## 📚 Documentation

In order to enable version-specific API documentation for napari, we need a place to host and archive it. Previously, this was done on napari/napari.github.io, and the proposal here is to enable that again - but this may change the way the docs are deployed.

As of now, this is the process:
- A PR is merged to the napari/napari `main` branch;
- The [Build and Deploy docs](https://github.com/napari/napari/blob/main/.github/workflows/deploy_docs.yml) action is triggered, and pushes the built html artifacts to the gh-pages branch of napari/napari.github.io, **overwriting** whatever content is already there.

One way we could orchestrate this process is the following:
- A PR is merged to the napari/napari `main` branch;
- The [Build and Deploy docs](https://github.com/napari/napari/blob/main/.github/workflows/deploy_docs.yml) action is triggered, and pushes the built html artifacts to the `dev` subfolder of the gh-pages branch of napari/napari.github.io, keeping the existing deployments for other versions.

This of course implies changing the subfolder structure of the current documentation (released versions would show up under `napari.org/<version>`, the latest release under `napari.org/stable`, and the current version under development would show under `napari.org/dev` or something similar.) This could mean a few broken links especially if people are linking from other pages to the napari docs.

Another change this would imply is an extra action item on releases, to set up the symlink from stable to the newest release.

cc @jni @neuromusic @codemonkey800

### PR DESCRIPTION
**Pull request by melissawm**
# Description
This PR adds a version switcher dropdown to the docs, and a github action to deploy these docs to the `napari/napari.github.io` upon merge to main (EDIT: This is ready)

This should be merged first, followed by https://github.com/napari/napari.github.io/pull/350

The caveat here, as mentioned in #4593 , is that links to the docs pages will change due to the new folder structure pointing to `/dev`, `/stable` etc. 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References
Closes #4593

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by melissawm** [comment]
A remaining question is if we want to use tags for the folder names, as suggested here https://github.com/napari/napari/issues/4593#issuecomment-1141676349 (thanks, @lucyleeow !)

**Comment by lucyleeow** [comment]
> The caveat here, as mentioned in https://github.com/napari/napari/issues/4593 , is that links to the docs pages will change due to the new folder structure pointing to /dev, /stable etc.

Potentially adding an index.html file in root of gh-pages that re-directs to stable could work. E.g., for skan I redirected to stable:

```html
<meta http-equiv="refresh" content="0; url=https://jni.github.io/skan/stable" />
```

**Comment by lucyleeow** [comment]
@melissawm I just had a look at the versioned docs in your repo (looks great!): https://melissawm.github.io/napari.github.io/dev/

The switch to stable did not seem to work for me, it went to: https://napari.org/stable/ instead of https://melissawm.github.io/napari.github.io/stable/ as per `version_switcher.json`

My guess would be that the `version` key in `version_switcher.json` needs to match `version_match` in the `config.py` (I looked at the [`config.py`](https://github.com/pydata/pydata-sphinx-theme/blob/31d1d09cb703d4aea49898922c2f1344f548c094/docs/conf.py#L87-L99) and [`switcher.json`](https://github.com/pydata/pydata-sphinx-theme/blob/main/docs/_static/switcher.json) of pydata theme and I *think* that is what it is doing...)

**Comment by melissawm** [comment]
Oh yes sure - this is how it will look after merged so I just didn't change that link. But all else is working so I'll put in a complete version. Thanks for checking!

**Comment by melissawm** [comment]
I am investigating a redirect alternative so the links will not be broken.

**Comment by melissawm** [comment]
Ok - might be a rudimentary solution but it seems to work :) @jni can you take a look when you have the time?

Feel free to push to it if you want to move it forward, as I won't be able to respond for a few days.

**Comment by jni** [comment]
We discussed in the docs working group that this is ready to merge, pending updates to the stable version number. Love the redirect! ❤️‍🔥 Thanks @melissawm! Please ping me on Zulip when this is ready to merge!

**Comment by richaagarwal** [comment]
@jni @melissawm Does this still need code review/approval? Anything I can do to help coordinate next steps?

**Comment by jni** [comment]
I was about to get this started the other day when our linux and doc builds started failing, so I decided to wait. 😅 Now I *think* @melissawm needs to merge main again into napari/napari.github.io#350, then we can try again?

**Comment by melissawm** [comment]
Just rebased - should be good to go! (will update napari/napari.github.io#350 now)

**Review by jni — APPROVED** [review]

**Pull request merged**
